#+TITLE:        RNA-Seq Workflow
#+STARTUP:      align fold noindent hideblocks
#+OPTIONS:      H:3 num:nil toc:2 ':t *:t \n:nil ::t f:t |:t
#+OPTIONS:      author:nil creator:nil html-postamble:nil tasks:nil

* Typical workflow for NGS data

** Processing steps

1. DNA extraction from a sample
2. DNA sequencing
3. Alignment of sequencing reads to a reference genome
4. Basic exploratory data analysis
5. Identification of genomic variants (SNPs, small insertions and deletions)
6. Gene quantification (i.e., statistics on count data)

** Tools

- [[https://github.com/ucdavis-bioinformatics/sickle][sickle]], for QC of reads
- [[http://bowtie-bio.sourceforge.net/bowtie2/index.shtml][Bowtie2]], or [[https://ccb.jhu.edu/software/tophat/index.shtml][TopHat2]] which relies on Bowtie, now [[https://ccb.jhu.edu/software/hisat2/index.shtml][HISAT2]] (3) : Fasta -> SAM
- [[http://www.htslib.org/download/][SAMtools]] (4+5)
- [[https://www.htslib.org/doc/bcftools.html][BCFtools]] (not to be confounded with [[https://bedtools.readthedocs.io/en/latest/][bedtools]])
- [[https://github.com/vcflib/vcflib][vcflib]], to manipulate VCF files

*** TODO [0/4]
- [ ] [[http://cole-trapnell-lab.github.io/cufflinks/][cufflinks]], part of the Tuxedo toolsuite (Bowtie + TopHat + cufflinks)
- [ ] [[https://broadinstitute.github.io/picard/][Picard]], to manipulate HTS files
- [ ] "How many reads map onto exons, introns and intergenic regions" on [[https://www.biostars.org/p/131734/][Biostars]]
- [ ] [[http://clavius.bc.edu/~erik/CSHL-advanced-sequencing/freebayes-tutorial.htm][Calling variants with freebayes]]

** Tutorials

- Workflows: https://is.gd/eKFGC2 + https://is.gd/In1Zcn
- [[https://osca.bioconductor.org][Orchestrating Single-Cell Analysis with Bioconductor]]

** File Formats

- Fasta :: Nucleotide or aa, with a description line (the symbol =>= followed by the sequence /identifier/ and other (optional) information, also called /comment/)


- Fastq :: Like Fasta file (header = =@= followed by sequence ID) but with the corresponding quality scores on a separate line (after the =+= sign); quality value ranges from 0x21 (!, lowest) to 0x7e (~, highest). Like Fasta files, they can be compressed using =gzip=.


- SRA :: Sequence Read Archives, from NCBI. Can be converted to Fasta using =fastq-dump=:

#+BEGIN_SRC bash
$ fastq-dump --gzip file.sra
#+END_SRC

As an alternative, samtools can be used to generate a BAM file:

#+BEGIN_SRC bash
$ sam-dump SRR6960207.sra | samtools view -bS - > alignments/SRR6960207.bam
#+END_SRC

- SAM (BAM) :: Composed of a header (=@= lines) and alignments of the sequence against a reference genome. BAM files are compressed version of SAM files.

#+CAPTION: Sequence Alignment/Map Format Specification
#+NAME: fig:sam-format
[[~/_img/fig-sam-format.png]]

- VCF :: Variant Call Format (See [[http://www.1000genomes.org][1000 Genomes Project]])

A typical workflow consists in calling mpileup via samtools or bcftools (which now integrate the pileup option). The mpileup command automatically scans every position supported by an aligned read, computes all the possible genotypes supported by these reads, and then computes the probability that each of these genotypes is truly present in our sample.

#+BEGIN_SRC bash
$ bcftools mpileup -f NC_008253.fna reads_aligned.sorted.bam -o sim_variants.bcf
#+END_SRC

The vcflib tools allow to further manipulate VCF files.
