* Statistical packages

** R

- [[https://cran.r-project.org/web/packages/ape/index.html][ape]] ([[http://ape-package.ird.fr][www]])
- [[https://cran.r-project.org/web/packages/apTreeshape/index.html][apTreeshape]]
- [[https://cran.r-project.org/web/packages/phytools/index.html][phytools]] 
- [[https://cran.r-project.org/web/packages/phangorn/index.html][phangorn]]  
- [[https://cran.r-project.org/web/packages/adephylo/index.html][adephylo]]

** Python

- [[https://pypi.org/project/phylogeny/][phylogeny]]
- [[http://etetoolkit.org][ETE]] ([[https://www.r-bloggers.com/phylogenies-in-r-and-python/][example]])
- [[https://pypi.org/project/DendroPy/][dendropy]] ([[https://academic.oup.com/bioinformatics/article/26/12/1569/287181][article]])
- [[https://biopython.org/wiki/Phylo][Bio.phylo]]

* Computational molecular evolution

** Evolutionary models 

[[https://en.wikipedia.org/wiki/Models_of_DNA_evolution][wikipedia]]

[[http://www.cs.toronto.edu/~brudno/csc2417_15/evophylo.pdf][Molecular Evolution, Substitution Models, and Phylogenies]]

substitution of selectively advantageous genes for less advantageous ones (population genetics); molecular evolution is dominated by mutations that are neutral from the standpoint of natural selection.
gene substitution = a mutant allele of a gene replaces the predominant allele in a population
Models of evolution are based on the assumption that sequences from different species have descended from some ancestral gene in a common ancestral species. Thus, divergence between sequences is a result of speciation. These genes are essentially the same and are called orthologues. The assumption may not hold because of gene duplication. Genes which have diverged from a common ancestor by gene duplication are called paralogues. They are different genes in the same organism.

Jukes-Cantor Model: Poisson model for substitutions, extended to a continuous time (4-state) Markov process; the probability of a nucleotide $i$ changing into nucleotide $j$ in time $t$ is $P_{ij} = 0.25\big(1-\exp(-4\lambda t)\big)$. Kimura (1980) extended this model by allowing transitions and transversions with different rates.

See also: [[https://math.la.asu.edu/~jtaylor/teaching/Spring2015/APM504/APM504.html][Applied Probability and Stochastic Processes]] (3 Feb), [[http://treethinkers.org/jukes-cantor-model-of-dna-substitution/][Jukes Cantor Model of DNA substitution]], Probability Models for DNA Sequence Evolution by Richard Durrett, 2nd ed. (Springer, 2008).

NEXT
- http://www.iqtree.org/doc/Substitution-Models
- https://telliott99.blogspot.com/2010/02/models-of-sequence-evolution-1.html

11 probabilistic models available in =ape::dist.dna=.

- *raw:* proportion of sites that differ between each pair of sequences

    c *g* c *t* *c* c *t* g *a* c
    c *a* c *g* *t* c *c* g *g* c     => 5/10 = 0.5

- *JC69:* This model was developed by Jukes and Cantor (1969). It assumes that all substitutions (i.e. a change of a base by another one) have the same probability. This probability is the same for all sites along the DNA sequence. This last assumption can be relaxed by assuming that the substition rate varies among site following a gamma distribution which parameter must be given by the user. By default, no gamma correction is applied. Another assumption is that the base frequencies are balanced and thus equal to 0.25.

"It can be shown that the mean number of base differences at a single position on the mRNA, $\mu$, is related to the observed fraction of residues with single base differences, $p$, by the expression:
$$ \mu = \frac{3}{4} \ln \frac{3}{3-4 p}. $$"
-- Thomas H. Jukes, [[http://www.garfield.library.upenn.edu/classics1990/A1990CZ67100002.pdf][How Many Nudeotide Substitutions Actually Took Place?]]

- *K80:* The distance derived by cite:kimura-1980-simpl-method, sometimes referred to as Kimura's 2-parameters distance, has the same underlying assumptions than the Jukes-Cantor distance except that two kinds of substitutions are considered: transitions (A <-> G, C <-> T), and transversions (A <-> C, A <-> T, C <-> G, G <-> T). They are assumed to have different probabilities. A transition is the substitution of a purine (C, T) by another one, or the substitution of a pyrimidine (A, G) by another one. A transversion is the substitution of a purine by a pyrimidine, or vice-versa. Both transition and transversion rates are the same for all sites along the DNA sequence. Jin and Nei (1990) modified the Kimura model to allow for variation among sites following a gamma distribution. Like for the Jukes-Cantor model, the gamma parameter must be given by the user. By default, no gamma correction is applied.

See also: cite:kimura-1981-estim, cite:tajima-1984-estim-evolut

Ultrametric trees = triangle inequality + d(x,y) â‰¤ max(d(x,z), d(y,z)) (max linkage clustering \equiv d(x,y) := max(d(x,z), d(y,z)) while single linkage \equiv d(x,y) := min(d(x,z), d(y,z)))


=ape::DNAbin= : efficient encoding of a sequence (nucleotid or a.a.) using 1 byte (8 bits):

A T C G
1 0 0 1 1 0 0 0
        base known
          alignment gap
            character unknown
              unused

[[http://www.bioinformatics.org/sms2/iupac.html][IUPAC code]]:

| Nucleotide       | IUPAC |
|------------------+-------|
| A or G           | R     |
| A or C           | M     |
| A or T           | W     |
| G or C           | S     |
| G or T           | K     |
| C or T           | Y     |
| A or G or C      | V     |
| A or C or T      | H     |
| A or G or T      | D     |
| G or C or T      | B     |
| A or G or C or T | N     |
